### PR TITLE
Support for proxying Juju client/controller connections through the Kubernetes API

### PR DESCRIPTION
**Pull request by tlm**
## Support for proxying Juju client through k8s

When Juju clients want to talk with Juju controllers inside of k8s they require the controller to be publicly accessible. This commit introduce a change allowing clients to port-forward to the controller when a route able connection is not possible.

Main features of the PR:
  - k8s tunnel and proxier for the k8s api
  - Start of a wider proxying framework for developing other types of proxy's for Juju controller connections
  - Changes to the users controller.yaml file describing the proxy to use when making connections.

## Checklist

 - [ ] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 Yes, leaving for the moment for another PR.
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [x] Comments answer the question of why design decisions were made

## QA steps

QA Steps is starting a Juju bootstrap. Install a Kubernetes cluster that Juju has previously failed on. Examples such as minikube or kind.

Then bootstrap juju to one of these clusters. The end result should be a perfectly working Juju.

The local users controller.yaml file can be checked for the new proxying information.

## Documentation changes

** Coming Soon

### CODE REVIEW

File: caas/kubernetes/tunnel_test.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		kubernetes.TunnelKindPods,
+		"test",
+		"test-pod",
+		"wef")
+
+	if err := tun.ForwardPort(); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Fatal("because I can")
```
Comment (hpidcock) [inline caas/kubernetes/tunnel_test.go:28]: ?

File: caas/kubernetes/tunnel.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+func (t *Tunnel) Close() {
+	close(t.stopChan)
+}
+
+func (t *Tunnel) findSuitablePodForService() (*corev1.Pod, error) {
+	clientSet := kubernetes.New(t.client)
+	service, err := clientSet.CoreV1().Services(t.Namespace).
+		Get(context.TODO(), t.Target, meta.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NewNotFound(err, "can't find service "+t.Target)
```
Comment (hpidcock) [inline caas/kubernetes/tunnel.go:58]: errors.NotFoundf(err, "service %s", t.Target)

File: caas/kubernetes/tunnel.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	t.LocalPort = local
+
+	ports := []string{fmt.Sprintf("%s:%s", t.LocalPort, t.RemotePort)}
+
+	pf, err := portforward.New(dialer, ports, t.stopChan, t.readyChan, t.Out, t.Out)
+	if err != nil {
+		return err
+	}
+
+	errChan := make(chan error)
```
Comment (hpidcock) [inline caas/kubernetes/tunnel.go:122]: change to buffered channel of 1 to prevent goroutine leak

File: cmd/juju/commands/bootstrap.go [hunk 4]
```diff
@@ -1034,6 +1036,14 @@ func (c *bootstrapCommand) controllerDataRefresher(
 		return errors.NewNotValid(nil, "unexpected error happened, IAAS mode should have environs.Environ implemented.")
 	}
 
+	var proxier proxy.Proxier
+	if conInfo, ok := environ.(environs.ConnectorInfo); ok {
+		proxier, err = conInfo.ConnectionProxyInfo()
+		if !errors.IsNotFound(err) && err != nil {
```
Comment (hpidcock) [inline cmd/juju/commands/bootstrap.go:1042]: err != nil && !errors.IsNotFound(err)

File: jujuclient/proxy.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		return errors.Annotate(err, "unmarshalling raw proxy config")
+	}
+
+	maker, err := factory.MakerForTypeKey(proxyConf.Type)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err = proxyConf.Config.Decode(maker.Config()); err != nil {
+		return errors.Annotatef(err, "deconding config for proxy of type %s", proxyConf.Type)
```
Comment (hpidcock) [inline jujuclient/proxy.go:61]: deconding -> decoding

File: proxy/proxy.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package proxy
+
+type Proxier interface {
```
Comment (hpidcock) [inline proxy/proxy.go:6]: Comment all interface methods

File: caas/kubernetes/provider/proxy/info.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+func GetControllerProxy(
+	name,
+	apiHost string,
+	configI core.ConfigMapInterface,
+	saI core.ServiceAccountInterface,
+	secretI core.SecretInterface,
+) (*Proxier, error) {
+	cm, err := configI.Get(context.TODO(), name, meta.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("no controller proxy config found for %s", name)
```
Comment (hpidcock) [inline caas/kubernetes/provider/proxy/info.go:32]: change to "controller proxy config for %s", NotFoundf already has the not found text in the Error() string.

File: caas/kubernetes/provider/proxy/info.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	config := ControllerProxyConfig{}
+	if err := json.Unmarshal([]byte(cm.Data[ProxyConfigMapKey]), &config); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	sa, err := saI.Get(context.TODO(), config.Name, meta.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("no controller proxy service account found for %s", name)
```
Comment (hpidcock) [inline caas/kubernetes/provider/proxy/info.go:44]: change to "controller proxy service account for %s", NotFoundf already has the not found text in the Error() string.

File: caas/kubernetes/tunnel.go [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+			LabelSelector: labels.SelectorFromSet(service.Spec.Selector).String(),
+		})
+
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	podCount := len(pods.Items)
+	if podCount == 0 {
+		return nil, errors.NotFoundf("no pods founds for service %s", t.Target)
```
Comment (hpidcock) [inline caas/kubernetes/tunnel.go:74]: change to "pods for service %s", NotFoundf already has the not found text in the Error() string.

**Review by hpidcock — APPROVED** [review]
Great work, just a few little details and one bug.

File: caas/kubernetes/tunnel.go (hunk 2 above)
Comment (tlm) [reply caas/kubernetes/tunnel.go:58]: This function signature doesn't exist

File: caas/kubernetes/tunnel.go (hunk 3 above)
Comment (tlm) [reply caas/kubernetes/tunnel.go:122]: Done & Done

File: cmd/juju/commands/bootstrap.go (hunk 4 above)
Comment (tlm) [reply cmd/juju/commands/bootstrap.go:1042]: Done & Done

File: jujuclient/proxy.go (hunk 5 above)
Comment (tlm) [reply jujuclient/proxy.go:61]: Done & Done

File: caas/kubernetes/provider/proxy/info.go (hunk 7 above)
Comment (tlm) [reply caas/kubernetes/provider/proxy/info.go:32]: Done & Done

File: caas/kubernetes/provider/proxy/info.go (hunk 8 above)
Comment (tlm) [reply caas/kubernetes/provider/proxy/info.go:44]: Done & Done

File: caas/kubernetes/tunnel.go (hunk 9 above)
Comment (tlm) [reply caas/kubernetes/tunnel.go:74]: Done & Done

File: proxy/proxy.go (hunk 6 above)
Comment (tlm) [reply proxy/proxy.go:6]: Done & Done

**Comment by tlm** [comment]
$$merge$$

**Pull request merged**